File: leetcode/plus_one/plus_one.py

def plusOne(digits: list[int]) -> list[int]:
    index = len(digits) - 1
    digits[index] += 1

    while index >= 0:
        if digits[index] == 10:
            digits[index] = 0
            if index - 1 >= 0:
                digits[index - 1] += 1
            else:
                digits.insert(0, 1)
        index -= 1

    return digits


plusOne([1, 2, 3])
plusOne([4, 3, 2, 1])
plusOne([9])

File: leetcode/plus_one/plus_one.java
class Solution {
    public int[] plusOne(int[] digits) {
        int lastIndex = digits.length - 1;
        if (digits[lastIndex] != 9) {
            digits[lastIndex] += 1;
            return digits;
        }
        
        int carryIn = 1;
        for (; lastIndex > 0; lastIndex--) {
            digits[lastIndex] += carryIn;
            if (digits[lastIndex] == 10) {
                digits[lastIndex] = 0;
                carryIn = 1;
            } else {
                carryIn = 0;
            }
        }
        digits[lastIndex] += carryIn;
        if (digits[lastIndex] == 10) {
            digits[lastIndex] = 0;
            int[] newDigits = new int[digits.length + 1];
            newDigits[0] = 1;
            for (int i = 1; i < newDigits.length; i++) {
                newDigits[i] = digits[i - 1];
            }
            return newDigits;
        }
        
        return digits;
    }
}
